lib/proxy/pluggable.rb

```diff
@@ -1,131 +1,129 @@
+    if apath.nil?
+      @settings_file || "#{plugin_name}.yml"
+    else
+      @settings = nil
+      @settings_file = apath
     end
+  end
 
-    def settings
-      self.class.settings
-    end
+  def default_settings(a_hash = {})
+    @settings = nil
+    @plugin_default_settings ||= {}
+    @plugin_default_settings.merge!(a_hash)
+  end
 
-    def log_used_default_settings
-      settings.defaults.select {|k,v| settings.used_defaults.include?(k)}.
-        inject({}) {|acc, c| acc[c[0].to_s] = c[1]; acc}.
-        sort.
-        collect {|c| ":#{c[0]}: #{c[1]}"}.
-        join(", ")
-    end
+  def initialize_after(*module_names)
+    raise "#{plugin_name}: 'initialize_after' method has been removed."
+  end
 
-    def after_activation
-      instance_eval(&self.class.after_activation_blk) if self.class.after_activation_blk
-    end
+  def validate_readable(*settings)
+    validate(*settings.push(:file_readable => true))
+  end
 
-    def validate!(all_plugins)
-      validate_dependencies!(all_plugins, self.class.dependencies)
-      execute_validators(self.class.plugin_default_settings.keys.map {|k| ::Proxy::PluginValidators::Presence.new(self.class, k)})
-      execute_validators(self.class.validators)
-    end
+  def validate_presence(*settings)
+    validate(*settings.push(:presence => true))
+  end
 
-    def execute_validators(validators)
-      validators.each { |validator| validator.validate! }
-    end
+  def validate(*settings)
+    validator_params = settings.pop
+    predicate = validator_params.delete(:if)
+    validator_name = validator_params.keys.first
+    validator_args = validator_params[validator_name]
 
-    def validate_dependencies!(all_plugins, dependencies)
-      dependencies.each do |dep|
-        plugin = all_plugins.find {|p| p[:name] == dep.name}
-        raise ::Proxy::PluginNotFound, "Plugin '#{dep.name}' required by plugin '#{plugin_name}' could not be found." unless plugin
-        unless ::Gem::Dependency.new('', dep.version).match?('', self.class.cleanup_version(plugin[:instance].version))
-          raise ::Proxy::PluginVersionMismatch, "Available version '#{plugin[:instance].version}' of plugin '#{dep.name}' doesn't match version '#{dep.version}' required by plugin '#{plugin_name}'"
-        end
-      end
-    end
+    settings.each {|setting| validations << {:name => validator_name, :predicate => predicate, :args => validator_args, :setting => setting} }
+  end
 
-    def loading_failed(message)
-      raise ::Proxy::PluginLoadingAborted, message
-    end
+  def load_validators(class_name_to_use)
+    @validator_loader_class = class_name_to_use
   end
 
-  module ClassMethods
-    attr_reader :plugin_name, :version, :after_activation_blk, :bundler_group_name
+  def load_dependency_injection_wirings(class_name_to_use)
+    @di_wirings_loader_class = class_name_to_use
+  end
 
-    # Methods below define DSL for defining plugins
+  def load_programmable_settings(class_name_to_use)
```

Comment:
> Not sure I understand you correctly -- how would you specify an instance of a class (or anything else other than a class) in these spots?

Yeah, it's not possible now because `.new` is called.  I was trying to understand why that was needed, when this could just accept something that responds to `#load_programmable_settings` and not call `.new` first.  It's just the instantiation of the class seems like a peculiar interface in Ruby, where just responding to methods is usually enough to implement an interface.

Rather than `@programmable_settings.new.load_programmable_settings`, I'd expect `@programmable_settings.load_programmable_settings`, then either change `load_programmable_settings` to a class method or pass an instance of the Proxy::Example::RuntimeConfiguration in here.